File: LeetCode/1568. Minimum Number of Days to Disconnect Island/solution.py
class Solution:
    def minDays(self, grid: List[List[int]]) -> int:
        n, m = len(grid), len(grid[0])
        dxy = [-1,0,1,0,-1]
        marks = []
        for i in range(n):
            for j in range(m):
                if grid[i][j]: marks.append((i,j))

        def bfs(r, c):
            q = [(r,c)]
            visited = set([(r,c)])
            while q:
                nq = []
                for r, c in q:
                    for dr, dc in zip(dxy[:-1],dxy[1:]):
                        dr += r 
                        dc += c 
                        if 0 <= dr < n and 0 <= dc < m and grid[dr][dc] == 1 and (dr,dc) not in visited:
                            nq.append((dr,dc))
                            visited.add((dr,dc))
                q = nq 
            return visited

        if len(marks) <= 1:
            return len(marks)
        sr, sc = marks[0]
        seen = bfs(sr,sc)
        if len(seen) < len(marks):
            return 0
        for i, (r, c) in enumerate(marks):
            grid[r][c] = 0
            sr, sc = marks[i-1]
            seen = bfs(sr,sc)
            if len(seen) < len(marks)-1:
                return 1
            grid[r][c] = 1
        return 2


        

class Solution:
    def minDays(self, grid: List[List[int]]) -> int:
        n, m = len(grid), len(grid[0])
        def components():
            parent = [i for i in range(n*m)]

            def find(cur):
                if parent[cur] != cur:
                    parent[cur] = find(parent[cur])
                return parent[cur]
            def union(left,right):
                rl, rr = find(left), find(right)
                if rl != rr:
                    parent[rl] = rr 
                return 

            for i in range(n):
                for j in range(m):
                    if grid[i][j] == 0: continue
                    for x, y in [(-1,0),(1,0),(0,-1),(0,1)]:
                        x += i
                        y += j
                        if 0 <= x < n and 0 <= y < m and grid[x][y] == 1:
                            union(x*m+y, i*m+j)
            return len(set([find(i*m+j) for i in range(n) for j in range(m) if grid[i][j] == 1]))

        if components() > 1:
            return 0
        for i in range(n):
            for j in range(m):
                if grid[i][j] == 1:
                    grid[i][j] = 0
                    if components() > 1:
                        return 1
                    grid[i][j] = 1
        return 2

File: LeetCode/1568. Minimum Number of Days to Disconnect Island/solution.cpp
vector<pair<int,int>> dxy = {{-1,0},{1,0},{0,-1},{0,1}};

class Solution {
public:
    int dfs(int i, int j, vector<vector<int>> &visited){
        int cnt = 1, n = visited.size(), m = visited[0].size();
        visited[i][j] = 0;
        for (auto [dx,dy] : dxy){
            dx += i;
            dy += j;
            if (dx >= 0 && dx < n && dy >= 0 && dy < m && visited[dx][dy] == 1)
                cnt += dfs(dx,dy,visited);
        }
        return cnt;
    }
    bool check(vector<vector<int>> &visited){
        int tot = 0, n = visited.size(), m = visited[0].size();
        for (int i = 0; i < n; i++){
            for (int j = 0; j < m; j++){
                if (visited[i][j]){
                    if (tot)
                        return false;
                    tot += dfs(i,j,visited);
                }
            }
        }
        return tot>0;
    }
    int minDays(vector<vector<int>>& grid) {
        vector<vector<int>> visited = grid;
        int n = visited.size(), m = visited[0].size();
        if(!check(visited))
            return 0;
        for (int i = 0; i < n; i++){
            for (int j = 0; j < m; j++){
                if (grid[i][j] == 1){
                    visited = grid;
                    visited[i][j] = 0;
                    if (!check(visited))
                        return 1;
                }
            }
        }
        return 2;

    }
};














vector<pair<int,int>> dxy = {{-1,0},{1,0},{0,-1},{0,1}};

class Solution {
    vector<int> low, tin;
    int clock, n, m;
    bool found = false;
public:
    void tarjan(int cur, int parent, vector<vector<int>> &grid){
        int i = cur/m, j = cur%m, pi = parent/m, pj = parent%m;
        low[cur] = tin[cur] = clock++;
        for (auto [dx,dy] : dxy){
            dx += i;
            dy += j;
            if (dx < 0 || dx >= n || dy < 0 || dy >= m || grid[dx][dy] == 0)
                continue;
            int nxt = dx*m+dy;
            if (nxt == parent)
                continue;
            else if (tin[nxt] != -1){
                low[cur] = min(low[cur], tin[nxt]);
            }
            else{
                tarjan(nxt,cur,grid);
                low[cur] = min(low[cur], low[nxt]);
                if (low[nxt] > tin[cur])
                    found = true;
            }
        }
        // cout << i << " " << j << " " << child << endl;
        return;
    }
    int dfs(int i, int j, vector<vector<int>> &visited){
        int cnt = 1;
        visited[i][j] = 0;
        for (auto [dx,dy] : dxy){
            dx += i;
            dy += j;
            if (dx >= 0 && dx < n && dy >= 0 && dy < m && visited[dx][dy] == 1)
                cnt += dfs(dx,dy,visited);
        }
        return cnt;
    }
    bool check(vector<vector<int>> &visited){
        int tot = 0;
        for (int i = 0; i < n; i++){
            for (int j = 0; j < m; j++){
                if (visited[i][j]){
                    if (tot)
                        return false;
                    tot += dfs(i,j,visited);
                }
            }
        }
        return tot>0;
    }
    int minDays(vector<vector<int>>& grid) {
        vector<vector<int>> visited = grid;
        n = visited.size(), m = visited[0].size();
        if(!check(visited))
            return 0;
        clock = 0;
        tin.assign(n*m,-1);
        low.assign(n*m,-1);
        for (int i = 0; i < n; i++){
            for (int j = 0; j < m; j++){
                if (grid[i][j] == 1){
                    tarjan(i*m+j,-1,grid);
                    // cout << clock << endl;
                    if (clock == 1 || (clock > 2 && found))
                        return 1;
                    return 2;
                }
            }
        }
        return 2;

    }
};